### ISSUE TITLE
Make an example application using MapperCrawler

### ISSUE THREAD

**Issue by snorrwe** [snorrwe/lollygag#39]

**Comment by daringer** [comment]
will do it ;)

**Comment by snorrwe** [comment]
Excellent :)

**Comment by daringer** [comment]
well, seems like I've collected quite some issues on the way:
(1) found various issues in `utility/url.py`, added tons of unit-tests and fixed many corner-cases
(2) new helper `is_usable_link()`, because I had quite a lot of cases were the proto/domain/link split did not work out well (to sort out e.g., `mailto:`, anchor-only links and others)
(3) had ssl-warnings because of unverified certs, added a cmdline-arg to switch cert-verification on/off
(4) "-u" for adding urls to the cmdline seemed a little un intuitive to me, so now they are positional arguments, means the "-u" is not necessary, but I kept the "-u" anyways for legacy reasons 
(5) for the logging, I was a little surprised that in loglevel `info` most of the built-in logging is still spamming the screen, thus I changed some of the *built-in* logging to `debug()` in order to be able to use `info()` as the let's say *user-logging*
(6) for the `MapperCrawler` I am not aware why it's deriving from `DomainCrawler`, `Crawler` seems to be more convenient from my point-of-view. So various crawlers might implement different *crawling-features* and at the end one can simply mix together any number of them to get the desired feature combination. I've done some things towards this direction, would be very eager to hear your opinion. Anyways, `MapperCrawler` has been heavily modified to make it independent from the `DomainCrawler`, have an always consistent tree/graph structure (mapping) and  `max_depth` is used to constraint the data-collection in terms of tree/graph depth.

For sure the MapperCrawler-example is build based on that, but very likely you have been hoping for something else :smile:. I will try to separate the points above as good as possible and create specific issues, some are clearly bugs/helpers, others are let's say *design-ideas*. I'll let you know once the first commits are in...

**Comment by daringer** [comment]
My Foto contains "mapper", in Lüdinghausen all the Pull requests and additionally the mappercrawler including an example... Not Fully done but Workshop good until now

**Comment by snorrwe** [comment]
Yeah, I've seen them, I think I'll have time today to review the changes.
I've noticed that they did not pass CI checks. That will have to be investigated.

About (1): That's kinda embarrassing, thank you! 

About (4): I was thinking about getting rid of `argparse` altogether, because it seems odd for a library to hijack `argparse`. Maybe provide another `ConfigService` implementation without the `argparse` by default? What do you think?

About (6): You are absolutely right. The main reason was that originally I implemented `DomainCrawler` then `MapperCrawler`, then `Crawler`. Not switching `MapperCrawler's` base when `Crawler` was made available is definitely a design flaw. (So is not starting with `Crawler` in the first place).

**Comment by daringer** [comment]
Saw your comments, looks like minimal stuff, so shouldn't take long.

(4) I think actually argparse does a good job here for what it's needed for in this case, do not see the need to reinvent the parts argparse is actually solving pretty well here,

(6) Lucky me that I didn't go the wrong direction, very nice to hear. MapperCrawler looks ok right now, but still squashing, more testing and details have to be done. Will first do the current pull requests and then clean the MapperCrawler stuff up to and set-up another pull-request containing it...
